wordpress-mobile/WordPress-iOS
# Tools/code freeze script

This PR adds a script that handles a part of the Code Freeze process. 
The script:
1. performs some sanity checks;
2. creates the release branch;
3. set the branch properties to limit merges, commits and so;
4. bumps the versions in the project;
5. updates the file with the strings that have to be localised;
6. commits the changes and pushes to GitHub;
7. updates the GitHub milestone with the frozen tag;
8. generates the list of PRs merged in this release.

(Check the _Release Scenario_ for further details)

### To Test
Testing can be a bit tricky as some of these operations are to be done only during an actual code freeze. Anyway, all of them can be reverted, so it's possible to test it. Just remember to go back and delete all the changes.

1. Setup the GitHub access;
2. Put a fake release tag in the repository, tagging the same version that's in the project file;
3. run the script: `cd Scripts` and `bundle exec fastlane code_freeze`;
4. when asked, verify the it has picked the correct current version and answer `y`;

When it finishes, verify that:
1. there's a new `release/XX` branch in the repo, where XX is the version the script picked up, plus 1;
2. verify that the version in the `xcconfig`, `DeliverFile` and `download_metadata.swift` files has been updated;
3. verify that the new branch has been pushed to GitHub;
4. verify that the right protection rules are set for the branch;
5. verify that the related milestone has the frozen tag;
6. verify that the `~/wpios_prs_list_#{old_version}_#{new_version}.txt` file has been generated in your home with the list of merged PRs.

At the end of the test:
1. Remove branch protection from GitHub and delete the branch;
2. Remove the frozen tag from the GitHub milestone;
3. delete the local release branch.
4. delete the fake tag you added in point 2 of the setup.
